### PR TITLE
Transaction state machine, Buy modal, improved cancel hook

### PR DESCRIPTION
**Pull request by AlexanderKolberg**

### CODE REVIEW

File: playgrounds/react-vite/src/tabs/Collectable.tsx [hunk 1]
```diff
@@ -109,79 +111,138 @@ function Actions() {
...
+    </Box>
   );
 }
 
 function ListingsTable() {
   const { collectionAddress, chainId, collectibleId } = useMarketplace();
+  const [page, setPage] = useState(0);
+
+  const nextPage = () => setPage((prev) => prev + 1);
+  const prevPage = () => setPage((prev) => prev - 1);
```
Comment (w1that) [inline playgrounds/react-vite/src/tabs/Collectable.tsx:163]: For some reason, we won't able to click go back to previous page if we get the end of the list
<video src="https://github.com/user-attachments/assets/85dac4f1-bec6-465c-bbff-c7e442156339" />

File: playgrounds/react-vite/src/tabs/Collectable.tsx [hunk 2]
```diff
@@ -222,27 +290,47 @@ function OrdersTable({
...
-            onAction={onAction}
-          />
-        ))}
-      </TableBody>
-    </Table>
+    <>
+      <Table>
+        <TableHeader>
+          <TableRow>
+            <TableHead>Price</TableHead>
```
Comment (w1that) [inline playgrounds/react-vite/src/tabs/Collectable.tsx:297]: Apparently, we don't always get formatted price using as:
`//OrdersTableRow`
`<TableCell>{order.priceAmountFormatted}</TableCell>`

what about using `formatUnits` from viem instead?

Ekran Resmi 2024-11-28 12 02 19

File: playgrounds/react-vite/src/tabs/Collectable.tsx [hunk 3]
```diff
@@ -109,79 +111,138 @@ function Actions() {
...
 
 function ListingsTable() {
   const { collectionAddress, chainId, collectibleId } = useMarketplace();
+  const [page, setPage] = useState(0);
+
+  const nextPage = () => setPage((prev) => prev + 1);
+  const prevPage = () => setPage((prev) => prev - 1);
+
   const { data: listings, isLoading } = useListListingsForCollectible({
     collectionAddress,
```
Comment (w1that) [inline playgrounds/react-vite/src/tabs/Collectable.tsx:166]: I don't know if it's related for using this hook, but if no more page, it still shows next page:

https://github.com/user-attachments/assets/10634c09-8661-4b22-9bea-0e86ac10e04a

File: packages/sdk/src/react/ui/modals/BuyModal/_store.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import type { ShowBuyModalArgs } from ".";
+
+export interface BuyModalState {
+  isOpen: boolean;
+  open: (args: ShowBuyModalArgs) => void;
+  close: () => void;
+  state: {
+    collectionAddress: Hex;
+    chainId: string;
+    tokenId: string;
```
[the remark this answers is not in the document]
Comment (AlexanderKolberg) [reply packages/sdk/src/react/ui/modals/BuyModal/_store.ts:14]: Yeah, we can simplify this we only need the order and quantity

File: packages/sdk/src/react/ui/modals/BuyModal/index.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+};
+
+export const useBuyModal = () => {
+  return {
+    show: (args: ShowBuyModalArgs) => buyModal$.open(args),
+    close: () => buyModal$.close(),
+  };
+};
+
+export const BuyModal = () => {
```
Comment (w1that) [inline packages/sdk/src/react/ui/modals/BuyModal/index.tsx:25]: After closing the modal, same Buy CTA doesn't work to open it again, while clicking on another buy CTA opens. Btw, I'm not sure if we should use animation for this modal, if we will, we should add to others too. 
<video src="https://github.com/user-attachments/assets/ad898e7b-73a5-4f7e-bbf4-d0f67746ed8b" />

File: playgrounds/react-vite/src/tabs/Collectable.tsx (hunk 2 above)
Comment (AlexanderKolberg) [reply playgrounds/react-vite/src/tabs/Collectable.tsx:297]: Got backend to add a fix for this, will be deployed today

**Review by w1that — APPROVED** [review]

**Pull request merged**